Algorithm.Framework/Portfolio/SectorWeightingPortfolioConstructionModel.cs

```diff
@@ -0,0 +1,170 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect.Algorithm.Framework.Alphas;
+using QuantConnect.Data.Fundamental;
+using QuantConnect.Data.UniverseSelection;
+using QuantConnect.Securities;
+
+namespace QuantConnect.Algorithm.Framework.Portfolio
+{
+    /// <summary>
+    /// Provides an implementation of <see cref="IPortfolioConstructionModel"/> that generates percent targets based on the
+    /// <see cref="CompanyReference.IndustryTemplateCode"/>. 
+    /// The target percent holdings of each sector is 1/S where S is the number of sectors and
+    /// the target percent holdings of each security is 1/N where N is the number of securities of each sector.
+    /// For insights of direction <see cref="InsightDirection.Up"/>, long targets are returned and for insights of direction
+    /// <see cref="InsightDirection.Down"/>, short targets are returned.
+    /// It will ignore <see cref="Insight"/> that have no <see cref="CompanyReference.IndustryTemplateCode"/> value.
+    /// </summary>
+    public class SectorWeightingPortfolioConstructionModel : EqualWeightingPortfolioConstructionModel
+    {
+        private readonly Dictionary<Symbol, string> _sectorCodeBySymbol = new Dictionary<Symbol, string>();
+
+        /// <summary>
+        /// Initialize a new instance of <see cref="SectorWeightingPortfolioConstructionModel"/>
+        /// </summary>
+        /// <param name="rebalancingFunc">For a given algorithm UTC DateTime returns the next expected rebalance time</param>
+        public SectorWeightingPortfolioConstructionModel(Func<DateTime, DateTime> rebalancingFunc)
```

Comment:
Now that the API refactor PR is merged we should add the new constructor overloads taking `IDateRules/PyObject/Func<DateTime,DateTime?>`